File: LC/15_3Sum.py
# Sort the array
# Check the combinations
# When arriving into a > 0 val as starting value, end and return what was find

from typing import List


def ThreeSum(self, nums: List[int]) -> List[List[int]]:
	nums.sort()
	res = []
	for i, a in enumerate(nums):
		if i > 0 and a == nums[i - 1]:
			continue
		l, r = i + 1, len(nums) - 1
		while l < r:
			threeSum = a + nums[l] + nums[r]
			if threeSum > 0: 
				r -= 1
			elif threeSum < 0:
				l += 1
			else:
				res.append([a, nums[l], nums[r]])
				l += 1
				while nums[l] == nums[l - 1] and l < r:
					l += 1 
	return res

File: LC/15_3Sum.go
package lc

import "sort"

// After picking the first number, implment ordered two sum, using two pointers to find the solution.
// It's a O(N^2) solution in worst case scenario
func Threesum(nums []int) (res [][]int) {
	sort.Ints(nums)
	var sum int
	for i, v := range nums {
		if i > 0 && v == nums[i-1] {
			continue
		}
		l := i + 1
		r := len(nums) - 1
		for l < r {
			sum = v + nums[l] + nums[r]
			switch {
			case sum > 0:
				r--
			case sum < 0:
				l++
			default:
				res = append(res, []int{v, nums[l], nums[r]})
				r--
				for nums[l] == nums[l-1] && l < r {
					r--
				}
			}

		}
	}
	return res
}
